rslib/src/search/sqlwriter.rs

```diff
@@ -184,30 +184,79 @@ impl SqlWriter<'_> {
         Ok(())
     }
 
-    fn write_unqualified(&mut self, text: &str) {
+    fn write_unqualified(&mut self, text: &str, no_combining: bool) -> Result<()> {
+        let text = to_sql(text);
+        let text = if no_combining {
+            without_combining(&text)
+        } else {
+            text
+        };
         // implicitly wrap in %
-        let text = format!("%{}%", &to_sql(text));
+        let text = format!("%{}%", text);
         self.args.push(text);
-        write!(
-            self.sql,
-            "(n.sfld like ?{n} escape '\\' or n.flds like ?{n} escape '\\')",
-            n = self.args.len(),
-        )
-        .unwrap();
-    }
+        let arg_idx = self.args.len();
 
-    fn write_no_combining(&mut self, text: &str) {
-        let text = format!("%{}%", without_combining(&to_sql(text)));
-        self.args.push(text);
-        write!(
-            self.sql,
-            concat!(
-                "(coalesce(without_combining(cast(n.sfld as text)), n.sfld) like ?{n} escape '\\' ",
-                "or coalesce(without_combining(n.flds), n.flds) like ?{n} escape '\\')"
-            ),
-            n = self.args.len(),
-        )
-        .unwrap();
+        let sfld_expr = if no_combining {
+            "coalesce(without_combining(cast(n.sfld as text)), n.sfld)"
+        } else {
+            "n.sfld"
+        };
+        let flds_expr = if no_combining {
+            "coalesce(without_combining(n.flds), n.flds)"
+        } else {
+            "n.flds"
+        };
+
+        if let Some(field_indicies_by_notetype) = self.included_fields_by_notetype()? {
+            let field_idx_str = format!("' || ?{arg_idx} || '");
+            let other_idx_str = "%".to_string();
+
+            let notetype_clause =
+                |UnqualifiedSearchContext {
+                     mid,
+                     num_fields,
+                     sortf_excluded,
+                     fields,
+                 }: &UnqualifiedSearchContext|
+                 -> String {
```

Comment:
This may come down to personal preference, but I find destructuring in a closure argument makes the code a bit hard to read. How about the following?

```diff
diff --git a/rslib/src/search/sqlwriter.rs b/rslib/src/search/sqlwriter.rs
index 0d3a3fc9d..f5dc84718 100644
--- a/rslib/src/search/sqlwriter.rs
+++ b/rslib/src/search/sqlwriter.rs
@@ -211,38 +211,32 @@ impl SqlWriter<'_> {
             let field_idx_str = format!("' || ?{arg_idx} || '");
             let other_idx_str = "%".to_string();

-            let notetype_clause =
-                |UnqualifiedSearchContext {
-                     mid,
-                     num_fields,
-                     sortf_excluded,
-                     fields,
-                 }: &UnqualifiedSearchContext|
-                 -> String {
-                    let field_index_clause = |range: &Range<u32>| {
-                        let f = (0..*num_fields)
-                            .filter_map(|i| {
-                                if i as u32 == range.start {
-                                    Some(&field_idx_str)
-                                } else if range.contains(&(i as u32)) {
-                                    None
-                                } else {
-                                    Some(&other_idx_str)
-                                }
-                            })
-                            .join("\x1f");
-                        format!("{flds_expr} like '{f}' escape '\\'")
-                    };
-                    let mut all_field_clauses: Vec<String> =
-                        fields.iter().map(field_index_clause).collect();
-                    if !sortf_excluded {
-                        all_field_clauses.push(format!("{sfld_expr} like ?{arg_idx} escape '\\'"));
-                    }
-                    format!(
-                        "(n.mid = {mid} and ({all_field_clauses}))",
-                        all_field_clauses = all_field_clauses.join(" or ")
-                    )
+            let notetype_clause = |ctx: &UnqualifiedSearchContext| -> String {
+                let field_index_clause = |range: &Range<u32>| {
+                    let f = (0..ctx.num_fields)
+                        .filter_map(|i| {
+                            if i as u32 == range.start {
+                                Some(&field_idx_str)
+                            } else if range.contains(&(i as u32)) {
+                                None
+                            } else {
+                                Some(&other_idx_str)
+                            }
+                        })
+                        .join("\x1f");
+                    format!("{flds_expr} like '{f}' escape '\\'")
                 };
+                let mut all_field_clauses: Vec<String> =
+                    ctx.fields.iter().map(field_index_clause).collect();
+                if !ctx.sortf_excluded {
+                    all_field_clauses.push(format!("{sfld_expr} like ?{arg_idx} escape '\\'"));
+                }
+                format!(
+                    "(n.mid = {mid} and ({all_field_clauses}))",
+                    mid = ctx.mid,
+                    all_field_clauses = all_field_clauses.join(" or ")
+                )
+            };
             let all_notetype_clauses = field_indicies_by_notetype
                 .iter()
                 .map(notetype_clause)
```